### ISSUE TITLE
__qualname__ inconsistent on Python 2, returns a descriptor on the class instead of class name

### ISSUE THREAD

**Issue by jamadden** [cython/cython#2562]
Over in https://github.com/zopefoundation/zope.app.apidoc/issues/25 we noticed an issue with the `__qualname__` of functions (at least) on Python 2. Because it's implemented as a `PyGetSetDef`, when accessed on the class, it returns a `getset_descriptor` object instead of the class name.

```pycon
Python 2.7.15 (default, May  2 2018, 00:53:27)
[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from Cython.Compiler.Visitor import recursively_replace_node
>>> recursively_replace_node
<cyfunction recursively_replace_node at 0x10cd04950>
>>> recursively_replace_node.__class__
<type 'cython_function_or_method'>
>>> recursively_replace_node.__qualname__
'recursively_replace_node'
>>> recursively_replace_node.__class__.__qualname__
<attribute '__qualname__' of 'cython_function_or_method' objects>
```

That breaks code expecting a string.

Python 3 (at least 3.6, the only version I checked) doesn't seem to have this problem; presumably the built-in meaning of `__qualname__` is taking over? The results are consistent with other built-in functions by returning a string.

```pycon
Python 3.6.6 (default, Jun 28 2018, 05:43:53)
[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from Cython.Compiler.Visitor import recursively_replace_node
>>> recursively_replace_node
<cyfunction recursively_replace_node at 0x10e9a66c0>
>>> recursively_replace_node.__class__
<class 'cython_function_or_method'>
>>> recursively_replace_node.__qualname__
'recursively_replace_node'
>>> recursively_replace_node.__class__.__qualname__
'cython_function_or_method'
>>> import os
>>> os.chmod.__class__.__qualname__
'builtin_function_or_method'
```

It seems like it would be a lot of work to transform the `PyGetSetDef` for `__qualname__` into a custom descriptor that does in the `__dict__` (the only way I know how to fix it, but I am no CPython API expert!). Python 2 functions don't generally have `__qualname__`, perhaps the easiest thing to do would just be to `#ifdef` the `__qualname__` out of the `__pyx_CyFunction_getsets`?

(I understand if that's not appealing either; this isn't a big deal, we plan to workaround it on that project.)

**Comment by scoder** [comment]
Isn't it just a matter of adding a constant string to the class dict? I think it's a bug that this used a descriptor in the first place. The [PEP](https://www.python.org/dev/peps/pep-3155/) only refers to the attribute on classes (and functions).

**Comment by jamadden** [comment]
> Isn't it just a matter of adding a constant string to the class dict?

Maybe? You're not supposed to be able to assign to `__qualname__`, though, so some code still needs to run to prevent that.

>  I think it's a bug that this used a descriptor in the first place.

That's how CPython implements it for its builtin functions, though. Compare the `__dict__` for `all` with the dict for a Cython function (Python 3 by necessity, since Python 2 doesn't have `__qualname__`):
```pycon
>>> type(all).__qualname__
'builtin_function_or_method'
>>> type(all).__dict__
mappingproxy({...
              '__ge__': <slot wrapper '__ge__' of 'builtin_function_or_method' objects>,
              '__module__': <member '__module__' of 'builtin_function_or_method' objects>,
              '__doc__': <attribute '__doc__' of 'builtin_function_or_method' objects>,
              '__name__': <attribute '__name__' of 'builtin_function_or_method' objects>,
              '__qualname__': <attribute '__qualname__' of 'builtin_function_or_method' objects>,
...
              '__text_signature__': <attribute '__text_signature__' of 'builtin_function_or_method' objects>})
>>> type(recursively_replace_node).__dict__
mappingproxy({
              '__module__': <member '__module__' of 'cython_function_or_method' objects>,
              'func_doc': <attribute 'func_doc' of 'cython_function_or_method' objects>,
              '__doc__': <attribute '__doc__' of 'cython_function_or_method' objects>,
              'func_name': <attribute 'func_name' of 'cython_function_or_method' objects>,
              '__name__': <attribute '__name__' of 'cython_function_or_method' objects>,
              '__qualname__': <attribute '__qualname__' of 'cython_function_or_method' objects>,
              '__self__': <attribute '__self__' of 'cython_function_or_method' objects>,
...
              '__annotations__': <attribute '__annotations__' of 'cython_function_or_method' objects>})

**Comment by jdemeyer** [comment]
Duplicate of #2772 and fixed by #2773.
